### PR TITLE
[BH-1515] Bedtime reminder notification issue

### PR DESCRIPTION
**Pull request by disonek**
**Description**

**Your checklist for this pull request**

Make sure that this PR:
- [x] Complies with our guidelines for contributions
- [x] Has unit tests if possible.
- [x] Has documentation updated

### CODE REVIEW

File: module-services/service-time/AlarmOperations.cpp [hunk 1]
```diff
@@ -338,6 +333,26 @@ namespace alarms
         }
     }
 
+    bool AlarmOperationsCommon::turnOffAlarmIfFoundInOngoingEvents(std::uint32_t id)
```
Comment (dawidwojtas-mudita) [inline module-services/service-time/AlarmOperations.cpp:336]: `const`

File: module-services/service-time/AlarmOperations.cpp [hunk 2]
```diff
@@ -338,6 +333,26 @@ namespace alarms
...
+
+        switchAlarmExecution(*(*found), false);
+        ongoingSingleEvents.erase(found);
+        processOngoingEvents();
+        handleActiveAlarmsCountChange();
+
+        return true;
+    }
+
+    bool AlarmOperationsCommon::turnOffAlarmIfFoundInBedtime(const uint32_t id)
```
Comment (dawidwojtas-mudita) [inline module-services/service-time/AlarmOperations.cpp:351]: It'd be better to use `const std::uint32_t`

File: products/BellHybrid/services/time/AlarmOperations.cpp [hunk 3]
```diff
@@ -152,6 +152,18 @@ namespace alarms
           bedtime(std::move(bedtimeSettingsProvider))
     {}
 
+    bool AlarmOperations::turnOffAlarmIfFoundInBedtime(const uint32_t id)
```
Comment (dawidwojtas-mudita) [inline products/BellHybrid/services/time/AlarmOperations.cpp:155]: `std::`

File: products/BellHybrid/apps/common/src/popups/BedtimeNotificationWindow.cpp [hunk 4]
```diff
@@ -42,6 +44,20 @@ namespace gui
...
+
+    bool BedtimeNotificationWindow::handleSwitchData(SwitchData *data)
+    {
+        if (data == nullptr) {
+            LOG_ERROR("Received null pointer");
+            return false;
+        }
+
+        auto *eventData    = dynamic_cast<gui::BedtimeReminderPopupRequestParams *>(data);
+        currentEventRecord = eventData->eventRecord;
```
[the remark this answers is not in the document]
Comment (disonek) [reply products/BellHybrid/apps/common/src/popups/BedtimeNotificationWindow.cpp:56]: I will fix it

File: products/BellHybrid/services/time/AlarmOperations.cpp (hunk 3 above)
Comment (disonek) [reply products/BellHybrid/services/time/AlarmOperations.cpp:155]: I will fix it

File: products/BellHybrid/services/time/include/time/AlarmOperations.hpp [hunk 5]
```diff
@@ -80,10 +80,18 @@ namespace alarms
...
         auto decide(TimePoint now) -> bool;
+        auto createEvent() -> std::shared_ptr<AlarmEventRecord>;
+        void pushEvent(const std::shared_ptr<AlarmEventRecord> &bedtimeEvent);
+        bool isEventInContainer(uint32_t id);
+        EventsContainer<SingleEventRecord>::iterator findEvent(uint32_t id);
+        void removeAllEvents();
 
       private:
         auto isTimeForBed(const TimePoint &now, const time_t &bedtime) -> bool;
+        constexpr static auto resonableIdOffset = 1000;
```
[the remark this answers is not in the document]
Comment (disonek) [reply products/BellHybrid/services/time/include/time/AlarmOperations.hpp:91]: I did copy it from another class :P  I think it works the same, but I will flip these paramters

File: products/BellHybrid/services/time/include/time/AlarmOperations.hpp [hunk 6]
```diff
@@ -80,10 +80,18 @@ namespace alarms
       public:
         explicit Bedtime(std::unique_ptr<AbstractBedtimeSettingsProvider> &&settingsProvider);
         auto decide(TimePoint now) -> bool;
+        auto createEvent() -> std::shared_ptr<AlarmEventRecord>;
+        void pushEvent(const std::shared_ptr<AlarmEventRecord> &bedtimeEvent);
+        bool isEventInContainer(uint32_t id);
+        EventsContainer<SingleEventRecord>::iterator findEvent(uint32_t id);
```
[the remark this answers is not in the document]
Comment (disonek) [reply products/BellHybrid/services/time/include/time/AlarmOperations.hpp:86]: If function parameters is not a redirection(reference or pointer) so parameters is copied or moved I do prefer do not use const

File: products/BellHybrid/services/time/AlarmOperations.cpp [hunk 7]
```diff
@@ -336,6 +352,40 @@ namespace alarms
...
+
+    bool Bedtime::isEventInContainer(uint32_t id)
+    {
+        return findEvent(id) == bedtimeSingleEvents.end() ? false : true;
+    }
+
+    EventsContainer<SingleEventRecord>::iterator Bedtime::findEvent(uint32_t id)
+    {
+        return findSingleEventById(bedtimeSingleEvents, id);
+    }
```
[the remark this answers is not in the document]
Comment (disonek) [reply products/BellHybrid/services/time/AlarmOperations.cpp:381]: If function parameters is not a redirection(reference or pointer) so parameters is copied or moved I do prefer do not use const

File: module-services/service-time/AlarmOperations.cpp (hunk 1 above)
Comment (disonek) [reply module-services/service-time/AlarmOperations.cpp:336]: If function parameters is not a redirection(reference or pointer) so parameters is copied or moved I do prefer do not use const

File: module-services/service-time/AlarmOperations.cpp (hunk 2 above)
Comment (disonek) [reply module-services/service-time/AlarmOperations.cpp:351]: If function parameters is not a redirection(reference or pointer) so parameters is copied or moved I do prefer do not use const

File: products/BellHybrid/services/time/AlarmOperations.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (swichu91) [reply products/BellHybrid/services/time/AlarmOperations.cpp]: It's true in the case of headers. For implementation, I can't see a reason not to use const.

File: module-services/service-time/AlarmOperations.cpp (hunk 1 above)
Comment (swichu91) [reply module-services/service-time/AlarmOperations.cpp:336]: It's true in the case of headers. For implementation, I can't see a reason not to use const.

File: module-services/service-time/AlarmOperations.cpp (hunk 2 above)
Comment (swichu91) [reply module-services/service-time/AlarmOperations.cpp]: It's true in the case of headers. For implementation, I can't see a reason not to use const.

File: products/BellHybrid/alarms/include/AbstractAlarmAction.hpp [hunk 8]
```diff
@@ -1,15 +1,18 @@
-// Copyright (c) 2017-2021, Mudita Sp. z.o.o. All rights reserved.
+// Copyright (c) 2017-2022, Mudita Sp. z.o.o. All rights reserved.
 // For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
 
 #pragma once
 
+#include <cstdint>
+#include <Tables/Record.hpp>
```
[the remark this answers is not in the document]
Comment (swichu91) [reply products/BellHybrid/alarms/include/AbstractAlarmAction.hpp:7]: For reference please check: `doc/MuditaCppCodingGuidelines.md`

**Review by swichu91 — APPROVED** [review]

**Review by dawidwojtas-mudita — APPROVED** [review]

File: products/BellHybrid/alarms/include/AbstractAlarmAction.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (disonek) [reply products/BellHybrid/alarms/include/AbstractAlarmAction.hpp:7]: will be fixed

File: module-services/service-time/AlarmOperations.cpp [hunk 9]
```diff
@@ -128,18 +128,13 @@ namespace alarms
 
     void AlarmOperationsCommon::turnOffRingingAlarm(const std::uint32_t id, OnTurnOffRingingAlarm callback)
     {
-        auto found = findSingleEventById(ongoingSingleEvents, id);
-        if (found == ongoingSingleEvents.end()) {
-            LOG_ERROR("Trying to turn off nonexisting event");
+        bool ongoingEventsResult = turnOffAlarmIfFoundInOngoingEvents(id);
+        bool bedtimeEventsResult = turnOffAlarmIfFoundInBedtime(id);
+        if (ongoingEventsResult == false and bedtimeEventsResult == false) {
```
[the remark this answers is not in the document]
Comment (disonek) [reply module-services/service-time/AlarmOperations.cpp:133]: will be fixed

File: module-services/service-time/AlarmOperations.hpp [hunk 10]
```diff
@@ -114,14 +116,17 @@ namespace alarms
...
+        EventsContainer<SingleEventRecord> nextSingleEvents;
+        EventsContainer<SingleEventRecord> ongoingSingleEvents;
+        EventsContainer<SnoozedAlarmEventRecord> snoozedSingleEvents;
 
         alarms::AlarmType getAlarmEventType(const SingleEventRecord &event);
         virtual void handleAlarmEvent(const std::shared_ptr<AlarmEventRecord> &event,
                                       alarms::AlarmType alarmType,
                                       bool newStateOn);
+        virtual bool turnOffAlarmIfFoundInBedtime(const std::uint32_t id);
+        bool turnOffAlarmIfFoundInOngoingEvents(const std::uint32_t id);
```
[the remark this answers is not in the document]
Comment (disonek) [reply module-services/service-time/AlarmOperations.hpp:128]: will be fixed

**Pull request merged**